lib/pbench/server/auth/__init__.py

```diff
@@ -0,0 +1,600 @@
+        The introspected JWT token contains the claims specified in
+        https://tools.ietf.org/html/rfc7662
+        Note: this is not supposed to be used in production, instead rely on
+              offline token validation because of security concerns mentioned in
+              https://www.rfc-editor.org/rfc/rfc7662.html#section-4
+        Args:
+            token: token value to introspect
+            token_info_uri: token introspection uri
+        Returns:
+            Extracted token information
+            {
+                "aud": "client_id",
+                "email_verified": "true",
+                "expires_in": "3572", # seconds
+                "access_type": "offline",
+                "exp": "1660087224",
+                "azp": "client_id",
+                "scope": "openid email profile",
+                "email": "user@example.com",
+                "sub": "106203657987367771589"
+            }
+        """
+        payload = {
+            "client_id": self.client_id,
+            "client_secret": self.client_secret_key,
+            "token": token,
+        }
+        return self._post(token_info_uri, data=payload).json()
+
+    def token_introspect_offline(
+        self,
+        token: str,
+        key: str,
+        audience: str = "account",
+        algorithms: List[str] = ["RS256"],
+        **kwargs,
+    ) -> JSON:
+        """
+        Utility method to decode access/Id tokens using the public key provided
+        by the identity provider.
+        The introspected JWT token contains the claims specified in
+        https://tools.ietf.org/html/rfc7662
+        Args:
+            token: token value to introspect
+            key: client public key
+            audience: jwt token audience/client, who this token was intended for
+            algorithms: Algorithm with which this JWT token was encoded
+        Returns:
+            Extracted token information
+            {
+                "aud": "client_id",
+                "email_verified": "true",
+                "expires_in": "3572", # seconds
+                "access_type": "offline",
+                "exp": "1660087224",
+                "azp": "client_id",
+                "scope": "openid email profile",
+                "email": "user@example.com",
+                "sub": "106203657987367771589"
+            }
+        """
+        return jwt.decode(
+            token, key, algorithms=algorithms, audience=audience, **kwargs
+        )
+
+    def get_userinfo(self, token: str = None) -> JSON:
+        """
+        The userinfo endpoint returns standard claims about the authenticated
+        user, and is protected by a bearer token.
+        http://openid.net/specs/openid-connect-core-1_0.html#UserInfo
+        Args:
+            token: Valid token to extract the userinfo
+        Returns:
+            Userinfo payload
+            {
+                "family_name": <surname>,
+                "sub": "106203657987367771589",
```

Comment:
Isn't this a "placeholder" of some sort? An angle-brace token would tell us what it *means* rather than just showing a number. (Unless it really is a constant enforced by the API...)